### PR TITLE
watchpoint verbosity control

### PR DESCRIPTION
**Pull request by kpgriesser**
The PR provides an initial mechanism to control the verbosity of watchpoints ( and anything else we want to control ) from the debug console. It's important to note that this is not thread safe and that any messages that are printed on the fly during simulation ( e.g. checking for triggers ) is going to be problematic for multi-thread/rank simulations. However, I think we we encapsulate the printing in the `msg` function we can put in an SST::Output object that can also print the rank and thread information conveniently.

```
> help verbose
verbose [mask]: set verbosity mask or print if no mask specified
        A mask is used to select which features to enable verbosity.
        To turn on all features set the mask to 0xffffffff
                16: Show trigger details

> w current_learning_rate_ changed
Added watchpoint #0
> r
NNBatchController[batch_controller:initTraining:1000]: Starting training phase
NNBatchController[batch_controller:initTraining:1000]: ### Training setup
NNBatchController[batch_controller:initTraining:1000]: epochs=10
NNBatchController[batch_controller:initTraining:1000]: X.rows()=20000
NNBatchController[batch_controller:initTraining:1000]: batch_size=128
NNBatchController[batch_controller:initTraining:1000]: train_steps=157
Entering interactive mode at time 24025000 
  WP0: AC : loss/optimizer/current_learning_rate_ ...

> verbose 0
verbose=0x0

> run
Entering interactive mode at time 40041000 
  WP0: AC : loss/optimizer/current_learning_rate_ ...

> verbose 1
verbose=0x1
> r
Entering interactive mode at time 56057000 
  WP0: AC : loss/optimizer/current_learning_rate_ ...

> v 16
verbose=0x10
> r
  Before Event Handler

    WatchPoint loss/optimizer/current_learning_rate_ tests:
      loss/optimizer/current_learning_rate_ CHANGED  -> false

    No trace buffer

  After Event Handler

    WatchPoint loss/optimizer/current_learning_rate_ tests:
      loss/optimizer/current_learning_rate_ CHANGED  -> false

    No trace buffer

  Before Clock Handler

    WatchPoint loss/optimizer/current_learning_rate_ tests:
      loss/optimizer/current_learning_rate_ CHANGED  -> false

    No trace buffer

  After Clock Handler

    WatchPoint loss/optimizer/current_learning_rate_ tests:
      loss/optimizer/current_learning_rate_ CHANGED  -> true

    No trace buffer

    SetInteractive

Entering interactive mode at time 72073000 
  WP0: AC : loss/optimizer/current_learning_rate_ ...
```

( will go back and remove extra \n characters )

More notes in the code review comments

### CODE REVIEW

File: src/sst/core/impl/interactive/simpleDebug.cc [hunk 1]
```diff
@@ -105,7 +105,7 @@ SimpleDebugger::SimpleDebugger(Params& params) :
         { "verbose", "[mask]: set verbosity mask or print if no mask specified\n"
                 "\tA mask is used to select which features to enable verbosity.\n"
                 "\tTo turn on all features set the mask to 0xffffffff\n"
-                "\t\t0x10: Show trigger details" },
+                "\t\t16: Show trigger details" },
```
Comment (kpgriesser) [inline src/sst/core/impl/interactive/simpleDebug.cc:108]: Recommending using decimal as the keyboard auto-complete code gets corrupted when typing '0x...' .
I'd like to use a mask instead of a level so that different debug objects can be controlled individually.

File: src/sst/core/impl/interactive/simpleDebug.cc [hunk 2]
```diff
@@ -1210,6 +1217,8 @@ SimpleDebugger::cmd_watch(std::vector<std::string>& tokens)
             return;
         }
         else {
+            // Every action gets the same verbosity as the console object
```
Comment (kpgriesser) [inline src/sst/core/impl/interactive/simpleDebug.cc:1220]: I think it would be better to enforce this in the constructor.
Having said that, it might be we move to a singleton or other design patter in the future so leaving as-is.

File: src/sst/core/impl/interactive/simpleDebug.cc [hunk 3]
```diff
@@ -1756,9 +1765,9 @@ CommandHistoryBuffer::searchAny(const std::string& s, std::string& newcmd)
 
 
 void
-SimpleDebugger::print_verbose(verbosityMask mask, std::string message)
+SimpleDebugger::msg(VERBOSITY_MASK mask, std::string message)
 {
-    if (! static_cast<uint32_t>(mask) && verbosityLevel ) return;
+    if (! static_cast<uint32_t>(mask) && verbosity ) return;
```
Comment (kpgriesser) [inline src/sst/core/impl/interactive/simpleDebug.cc]: here is where we can encapsulate other output mechanisms without changing the rest of the code.

File: src/sst/core/impl/interactive/simpleDebug.h [hunk 4]
```diff
@@ -40,8 +40,8 @@ const std::map<ConsoleCommandGroup, std::string> GroupText {
     { ConsoleCommandGroup::MISC, "Misc" },
 };
 
-enum class verbosityMask : uint32_t {
-    SHOW_TRIGGERS = 0b0001'0000 // 0x10
+enum class VERBOSITY_MASK : uint32_t {
```
Comment (kpgriesser) [inline src/sst/core/impl/interactive/simpleDebug.h:43]: would like to enforce this. Probably should move this to a common header. ( see watchpoint code where I just us 0x10 and add a comment )

File: src/sst/core/serialization/objectMap.h [hunk 5]
```diff
@@ -132,7 +132,7 @@ class ObjectMapComparison
 
     virtual bool        compare()         = 0;
     virtual std::string getCurrentValue() = 0;
-    virtual void        print()           = 0;
+    virtual void        print(std::ostream& stream) = 0;
```
Comment (kpgriesser) [inline src/sst/core/serialization/objectMap.h:135]: follows common design pattern used elsewhere in sst

File: src/sst/core/watchPoint.h [hunk 6]
```diff
@@ -48,6 +50,13 @@ class WatchPoint : public Clock::HandlerBase::AttachPoint, public Event::Handler
         virtual ~WPAction() = default;
         virtual std::string actionToString()             = 0;
         virtual void        invokeAction(WatchPoint* wp) = 0;
+        inline void setVerbosity(uint32_t v) { verbosity = v; }
+    protected:
+        uint32_t verbosity = 0;
```
Comment (kpgriesser) [inline src/sst/core/watchPoint.h:55]: if verbosity changes we need to update all the objects. Maybe through the command line console we can specify a custom verbosity for each object as well in which case this becomes more complicated.

File: src/sst/core/watchPoint.h [hunk 7]
```diff
@@ -144,6 +153,12 @@ class WatchPoint : public Clock::HandlerBase::AttachPoint, public Event::Handler
         ALL          = 15
     };
 
+    //TODO can we avoid code duplication in WatchPoint and the WatchPointAction?
```
Comment (kpgriesser) [inline src/sst/core/watchPoint.h:156]: It is annoying to put the same code in WatchPoint and in the inner classes. but I guess it's not that much code.

File: src/sst/core/impl/interactive/simpleDebug.cc (hunk 3 above)
Comment (skuntz) [reply src/sst/core/impl/interactive/simpleDebug.cc]: Should this be bitwise &

File: src/sst/core/watchPoint.cc [hunk 8]
```diff
@@ -12,18 +12,18 @@
 #include "sst_config.h"
 
 #include "sst/core/watchPoint.h"
-
 #include "sst/core/output.h"
 #include "sst/core/simulation_impl.h"
 #include "sst/core/sst_types.h"
 
 #include <string>
+#include "watchPoint.h"
```
Comment (skuntz) [inline src/sst/core/watchPoint.cc]: How is this different from the sst/core/watchPoint.h above?

**Review by skuntz — APPROVED** [review]
